src/vs/editor/contrib/stickyScroll/browser/stickyScrollController.ts

```diff
@@ -3,63 +3,281 @@
 import { EditorOption, RenderLineNumbersType } from 'vs/editor/common/config/editorOptions';
 import { StickyScrollWidget, StickyScrollWidgetState } from './stickyScrollWidget';
-import { StickyLineCandidateProvider, StickyRange } from './stickyScrollProvider';
+import { IStickyLineCandidateProvider, StickyLineCandidateProvider, StickyRange } from './stickyScrollProvider';
 import { IModelTokensChangedEvent } from 'vs/editor/common/textModelEvents';
 import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
-import * as dom from 'vs/base/browser/dom';
 import { IContextMenuService } from 'vs/platform/contextview/browser/contextView';
 import { MenuId } from 'vs/platform/actions/common/actions';
+import { IContextKey, IContextKeyService } from 'vs/platform/contextkey/common/contextkey';
+import { EditorContextKeys } from 'vs/editor/common/editorContextKeys';
+import { ClickLinkGesture } from 'vs/editor/contrib/gotoSymbol/browser/link/clickLinkGesture';
+import { IRange, Range } from 'vs/editor/common/core/range';
+import { getDefinitionsAtPosition } from 'vs/editor/contrib/gotoSymbol/browser/goToSymbol';
+import { goToDefinitionWithLocation } from 'vs/editor/contrib/inlayHints/browser/inlayHintsLocations';
+import { Position } from 'vs/editor/common/core/position';
+import { CancellationTokenSource } from 'vs/base/common/cancellation';
+import * as dom from 'vs/base/browser/dom';
 
-export class StickyScrollController extends Disposable implements IEditorContribution {
+interface CustomMouseEvent {
+	detail: string;
+	element: HTMLElement;
+}
+
+export interface IStickyScrollController {
+	get stickyScrollCandidateProvider(): IStickyLineCandidateProvider;
+	get stickyScrollWidgetState(): StickyScrollWidgetState;
+	focus(): void;
+	focusNext(): void;
+	focusPrevious(): void;
+	goToFocused(): void;
+	findScrollWidgetState(): StickyScrollWidgetState;
+	dispose(): void;
+}
+
+export class StickyScrollController extends Disposable implements IEditorContribution, IStickyScrollController {
 
 	static readonly ID = 'store.contrib.stickyScrollController';
 
 	private readonly _stickyScrollWidget: StickyScrollWidget;
-	private readonly _stickyLineCandidateProvider: StickyLineCandidateProvider;
+	private readonly _stickyLineCandidateProvider: IStickyLineCandidateProvider;
 	private readonly _sessionStore: DisposableStore = new DisposableStore();
 
 	private _widgetState: StickyScrollWidgetState;
 	private _maxStickyLines: number = Number.MAX_SAFE_INTEGER;
 
+	private _stickyElements: HTMLCollection | undefined;
+	private _numberStickyElements: number | undefined;
+
+	private _stickyRangeProjectedOnEditor: IRange | undefined;
+	private _candidateDefinitionsLength: number = -1;
+
+	private _focused: boolean = false;
+	private _focusDisposableStore: DisposableStore | undefined;
+	private _focusedStickyElement: HTMLDivElement | undefined;
+	private _focusedStickyElementIndex: number | undefined;
```

Comment:
Maybe it's easier to track the focused element only once, e.g only use the index and simply look up the nth element from the parent if you need it.

Reply:
Doing so causes errors, when the index is used on this._stickyElements which no longer contains that index after scrolling in the editor (the widget changes)